lib/Lower/Bridge.cpp

```diff
@@ -574,41 +380,328 @@ class FirConverter : public Fortran::lower::AbstractConverter {
-  void genFIR(const Fortran::parser::ChangeTeamConstruct &) { TODO(); }
-  void genFIR(const Fortran::parser::CriticalConstruct &) { TODO(); }
-  void genFIR(const Fortran::parser::DoConstruct &) { TODO(); }
-  void genFIR(const Fortran::parser::IfConstruct &) { TODO(); }
+    // Generate loop body code.
+    for (auto &e : *eval.evaluationList) {
+      genFIR(e, unstructuredContext);
+    }
+
+    // Generate end loop code.
+    if (infiniteLoop || whileCondition) {
+      genFIRUnconditionalBranch(doStmtEval.localBlocks[0]);
+    } else {
+      for (size_t i = incrementLoopInfo.size(); i > 0;) {
+        genFIRIncrementLoopEnd(incrementLoopInfo[--i]);
+      }
+    }
+  }
 
-  void genFIR(const SelectCaseConstruct &) { TODO(); }
-  void genFIR(const SelectRankConstruct &) { TODO(); }
-  void genFIR(const SelectTypeConstruct &) { TODO(); }
+  /// Generate FIR to begin a structured or unstructured increment loop.
+  void genFIRIncrementLoopBegin(IncrementLoopInfo &info) {
+    mlir::Location location = toLocation();
+    mlir::Type type = info.isStructured()
+                          ? mlir::IndexType::get(builder->getContext())
+                          : info.loopVariableType;
+    mlir::Value lowerValue = genFIRLoopIndex(info.lowerExpr, type);
+    mlir::Value upperValue = genFIRLoopIndex(info.upperExpr, type);
+    if (info.stepExpr.has_value()) {
+      info.stepValue = genFIRLoopIndex(*info.stepExpr, type);
+    }
+    info.loopVariable = createTemporary(location, *info.loopVariableSym);
+
+    // Structured loop - generate fir.loop.
+    if (info.isStructured()) {
+      info.insertionPoint = builder->saveInsertionPoint();
+      llvm::SmallVector<mlir::Value, 1> stepVector;
+      if (info.stepValue) {
+        stepVector.push_back(info.stepValue);
+      }
+      info.doLoop = builder->create<fir::LoopOp>(location, lowerValue,
+                                                 upperValue, stepVector);
+      builder->setInsertionPointToStart(info.doLoop.getBody());
+      mlir::Value shadow = builder->create<fir::ConvertOp>(
+          location, info.loopVariableType, info.doLoop.getInductionVar());
+      localSymbols.pushShadowSymbol(*info.loopVariableSym, shadow);
+      return;
+    }
 
-  void genFIR(const Fortran::parser::WhereConstruct &) { TODO(); }
+    // Unstructured loop preheader code - initialize tripVariable, loopVariable.
+    if (!info.stepValue) {
+      info.stepValue = builder->create<mlir::ConstantOp>(
+          location, builder->getIntegerAttr(info.loopVariableType, 1));
+    }
+    mlir::Value tripCount;
+    tripCount = builder->create<mlir::SubIOp>(location, upperValue, lowerValue);
+    tripCount =
+        builder->create<mlir::AddIOp>(location, tripCount, info.stepValue);
+    if (info.stepExpr.has_value()) {
+      tripCount = builder->create<mlir::SignedDivIOp>(location, tripCount,
+                                                      info.stepValue);
+    }
```

Comment:
Instead of testing if there is a step value everywhere, the code would be easier to understand if step value was set to a constant 1 when it isn't present. We can rely on the folders in the IR to reduce all the constant expressions.

Reply:
I'm somewhat inclined to disagree here.  In addition to the legibility of this code, there is also the legibility of the generated mlir code, which is improved for the usual default step case when this if is retained.

Let's bargain.  You give me a fir.loop create call that takes an mlir::Value step argument instead of the unfortunate vector of same, and I'll make this change.  (I'll probably remember.)  (I'm assuming that the vector argument is never actually useful, as it might just possibly be if a vector of size > 1 can be used for multiple dimension do concurrent or forall cases.)

Reply:
We'll just have to disagree.

The vector of length 0 is the way all the other MLIR Ops that I've seen pass optional arguments. (In fact, I've looked for and did not find tablegen syntax for optional.) So I'm strongly disinclined to go a different way for LoopOp.

I think the easy to read argument is very debatable. Instead of generating a single pattern regardless of how the LoopOp was constructed (which is tied to the surface syntax), we have different paths and increased cyclomatic complexity. That's for a pattern that is necessarily transitory and ought to be optimized away very quickly. That's what the passes in the optimizer are for after all.

Reply:
Ok.  Thanks for the additional discussion.  And I may just follow your preference at some point; I'll think about that.  Occasionally I decide later that I was wrong.